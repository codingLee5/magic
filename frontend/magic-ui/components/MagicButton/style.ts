import { createStyles } from "antd-style"
import type { CSSProperties } from "react"

export const useStyles = createStyles(
	(
		{ css, prefixCls, token, isDarkMode },
		{ justify }: { justify?: CSSProperties["justifyContent"] },
	) => ({
		magicButton: css`
			display: flex;
			align-items: center;
			justify-content: ${justify};
			box-shadow: none;

      .${prefixCls}-btn-icon { 
				display: flex;
				align-items: center;
				justify-content: center;
			}

			--${prefixCls}-button-default-hover-color: ${token.colorText} !important;
			--${prefixCls}-button-default-hover-border-color: ${token.colorBorder} !important;
			--${prefixCls}-button-default-hover-bg: ${token.magicColorUsages.fill[0]} !important;
			--${prefixCls}-button-default-bg: ${
			isDarkMode ? token.magicColorUsages.bg[1] : token.colorWhite
		} !important;
			--${prefixCls}-button-default-color: ${token.colorTextSecondary} !important;
		`,
	}),
)
