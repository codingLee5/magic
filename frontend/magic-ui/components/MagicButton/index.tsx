import { colorScales as baseColorScales } from "../ThemeProvider/colors"
import type { ButtonProps, ThemeConfig } from "antd"
import { Button, Tooltip } from "antd"
import type { GetAntdTheme } from "antd-style"
import { useTheme, ThemeProvider, cx } from "antd-style"
import { forwardRef, memo, useMemo, type CSSProperties, type ReactNode } from "react"
import { useStyles } from "./style"

export interface MagicButtonProps extends ButtonProps {
	justify?: CSSProperties["justifyContent"]
	theme?: boolean
	tip?: ReactNode
}

const MagicButton = memo(
	forwardRef<HTMLButtonElement, MagicButtonProps>(
		(
			{
				tip,
				className,
				justify = "center",
				theme = true,
				hidden,
				...props
			}: MagicButtonProps,
			ref,
		) => {
			const { styles } = useStyles({ justify })

			const { magicColorUsages } = useTheme()

			const themeConfigs = useMemo<ThemeConfig | GetAntdTheme | undefined>(() => {
				return theme
					? (appearence) => ({
							components: {
								Button: {
									colorLink: magicColorUsages.primary.default,
									colorLinkHover: magicColorUsages.primary.hover,
									colorLinkActive: magicColorUsages.primary.active,
									colorPrimary: magicColorUsages.primary.default,
									colorPrimaryHover: magicColorUsages.primary.hover,
									colorPrimaryActive: magicColorUsages.primary.active,
									textHoverBg:
										appearence === "dark"
											? baseColorScales.grey[8]
											: magicColorUsages.primaryLight.default,
								},
							},
					  })
					: undefined
			}, [
				magicColorUsages.primary.active,
				magicColorUsages.primary.default,
				magicColorUsages.primary.hover,
				magicColorUsages.primaryLight.default,
				theme,
			])

			if (hidden) {
				return null
			}

			return (
				<ThemeProvider theme={themeConfigs}>
					<Tooltip title={tip}>
						<Button
							ref={ref}
							className={cx(styles.magicButton, className)}
							{...props}
						/>
					</Tooltip>
				</ThemeProvider>
			)
		},
	),
)

export default MagicButton
