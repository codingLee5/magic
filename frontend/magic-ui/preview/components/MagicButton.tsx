import React from "react"
import { Space } from "antd"
import MagicButton from "../../components/MagicButton"
import ComponentDemo from "./Container"

const ButtonDemo: React.FC = () => {
	return (
		<div>
			<ComponentDemo
				title="按钮类型"
				description="支持多种按钮类型：默认、主要、虚线、链接、文本"
				code="type: 'default' | 'primary' | 'dashed' | 'link' | 'text'"
			>
				<Space wrap>
					<MagicButton>默认按钮</MagicButton>
					<MagicButton type="primary">主要按钮</MagicButton>
					<MagicButton type="dashed">虚线按钮</MagicButton>
					<MagicButton type="link">链接按钮</MagicButton>
					<MagicButton type="text">文本按钮</MagicButton>
				</Space>
			</ComponentDemo>

			<ComponentDemo
				title="按钮尺寸"
				description="支持三种尺寸：大、默认、小"
				code="size: 'large' | 'default' | 'small'"
			>
				<Space wrap>
					<MagicButton size="large">大按钮</MagicButton>
					<MagicButton>默认按钮</MagicButton>
					<MagicButton size="small">小按钮</MagicButton>
				</Space>
			</ComponentDemo>

			<ComponentDemo
				title="按钮状态"
				description="支持加载、禁用等状态"
				code="loading | disabled"
			>
				<Space wrap>
					<MagicButton loading>加载中</MagicButton>
					<MagicButton disabled>禁用按钮</MagicButton>
					<MagicButton type="primary" loading>
						主要加载
					</MagicButton>
				</Space>
			</ComponentDemo>

			<ComponentDemo
				title="按钮对齐"
				description="支持自定义内容对齐方式"
				code="justify: 'flex-start' | 'center' | 'flex-end'"
			>
				<Space direction="vertical" style={{ width: "100%" }}>
					<MagicButton justify="flex-start" style={{ width: 200 }}>
						左对齐按钮
					</MagicButton>
					<MagicButton justify="center" style={{ width: 200 }}>
						居中对齐按钮
					</MagicButton>
					<MagicButton justify="flex-end" style={{ width: 200 }}>
						右对齐按钮
					</MagicButton>
				</Space>
			</ComponentDemo>

			<ComponentDemo
				title="带图标的按钮"
				description="支持在按钮中添加图标"
				code="icon: ReactNode"
			>
				<Space wrap>
					<MagicButton icon="🔍">搜索</MagicButton>
					<MagicButton type="primary" icon="➕">
						添加
					</MagicButton>
					<MagicButton type="dashed" icon="📁">
						文件夹
					</MagicButton>
				</Space>
			</ComponentDemo>

			<ComponentDemo
				title="带提示的按钮"
				description="支持添加工具提示"
				code="tip: ReactNode"
			>
				<Space wrap>
					<MagicButton tip="这是一个提示">悬停查看提示</MagicButton>
					<MagicButton type="primary" tip="主要按钮提示">
						主要按钮
					</MagicButton>
				</Space>
			</ComponentDemo>
		</div>
	)
}

export default ButtonDemo
