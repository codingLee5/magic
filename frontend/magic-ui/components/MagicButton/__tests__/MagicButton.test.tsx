import { render, screen, fireEvent } from "@testing-library/react"
import { describe, it, expect, vi } from "vitest"
import MagicButton from "../index"
import MagicThemeProvider from "../../ThemeProvider"

const renderWithTheme = (component: React.ReactElement) => {
	return render(<MagicThemeProvider theme="light">{component}</MagicThemeProvider>)
}

describe("MagicButton", () => {
	it("应该正确渲染按钮", () => {
		renderWithTheme(<MagicButton>测试按钮</MagicButton>)
		expect(screen.getByRole("button", { name: "测试按钮" })).toBeInTheDocument()
	})

	it("应该支持不同的按钮类型", () => {
		const { rerender } = renderWithTheme(<MagicButton type="primary">主要按钮</MagicButton>)
		expect(screen.getByRole("button")).toHaveClass("magic-btn-primary")

		rerender(
			<MagicThemeProvider theme="light">
				<MagicButton type="dashed">虚线按钮</MagicButton>
			</MagicThemeProvider>,
		)
		expect(screen.getByRole("button")).toHaveClass("magic-btn-dashed")
	})

	it("应该支持点击事件", () => {
		const handleClick = vi.fn()
		renderWithTheme(<MagicButton onClick={handleClick}>点击按钮</MagicButton>)

		fireEvent.click(screen.getByRole("button"))
		expect(handleClick).toHaveBeenCalledTimes(1)
	})

	it("应该支持禁用状态", () => {
		renderWithTheme(<MagicButton disabled>禁用按钮</MagicButton>)
		expect(screen.getByRole("button")).toBeDisabled()
	})

	it("应该支持加载状态", () => {
		renderWithTheme(<MagicButton loading>加载按钮</MagicButton>)
		expect(screen.getByRole("button")).toHaveClass("magic-btn-loading")
	})

	it("应该支持自定义样式类名", () => {
		renderWithTheme(<MagicButton className="custom-class">自定义按钮</MagicButton>)
		expect(screen.getByRole("button")).toHaveClass("custom-class")
	})

	it("应该支持图标", () => {
		renderWithTheme(
			<MagicButton icon={<span data-testid="icon">🚀</span>}>带图标按钮</MagicButton>,
		)
		expect(screen.getByTestId("icon")).toBeInTheDocument()
	})

	it("应该支持 tooltip", () => {
		renderWithTheme(<MagicButton tip="这是一个提示">带提示按钮</MagicButton>)
		expect(screen.getByRole("button")).toBeInTheDocument()
	})

	it("当 hidden 为 true 时应该不渲染", () => {
		const { container } = renderWithTheme(<MagicButton hidden>隐藏按钮</MagicButton>)
		expect(container.firstChild).toBeNull()
	})

	it("应该支持不同的 justify 属性", () => {
		renderWithTheme(<MagicButton justify="flex-start">左对齐按钮</MagicButton>)
		const button = screen.getByRole("button")
		expect(button).toBeInTheDocument()
	})

	it("应该支持 ref 转发", () => {
		const ref = vi.fn()
		renderWithTheme(<MagicButton ref={ref}>Ref 按钮</MagicButton>)
		expect(ref).toHaveBeenCalled()
	})

	// 快照测试
	describe("快照测试", () => {
		it("默认按钮快照", () => {
			const { asFragment } = renderWithTheme(<MagicButton>默认按钮</MagicButton>)
			expect(asFragment()).toMatchSnapshot()
		})

		it("主要按钮快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicButton type="primary">主要按钮</MagicButton>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("虚线按钮快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicButton type="dashed">虚线按钮</MagicButton>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("文本按钮快照", () => {
			const { asFragment } = renderWithTheme(<MagicButton type="text">文本按钮</MagicButton>)
			expect(asFragment()).toMatchSnapshot()
		})

		it("链接按钮快照", () => {
			const { asFragment } = renderWithTheme(<MagicButton type="link">链接按钮</MagicButton>)
			expect(asFragment()).toMatchSnapshot()
		})

		it("禁用状态按钮快照", () => {
			const { asFragment } = renderWithTheme(<MagicButton disabled>禁用按钮</MagicButton>)
			expect(asFragment()).toMatchSnapshot()
		})

		it("加载状态按钮快照", () => {
			const { asFragment } = renderWithTheme(<MagicButton loading>加载按钮</MagicButton>)
			expect(asFragment()).toMatchSnapshot()
		})

		it("带图标按钮快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicButton icon={<span>🚀</span>}>带图标按钮</MagicButton>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("不同尺寸按钮快照", () => {
			const { asFragment: smallFragment } = renderWithTheme(
				<MagicButton size="small">小按钮</MagicButton>,
			)
			expect(smallFragment()).toMatchSnapshot()

			const { asFragment: largeFragment } = renderWithTheme(
				<MagicButton size="large">大按钮</MagicButton>,
			)
			expect(largeFragment()).toMatchSnapshot()
		})

		it("隐藏按钮快照", () => {
			const { asFragment } = renderWithTheme(<MagicButton hidden>隐藏按钮</MagicButton>)
			expect(asFragment()).toMatchSnapshot()
		})
	})
})
