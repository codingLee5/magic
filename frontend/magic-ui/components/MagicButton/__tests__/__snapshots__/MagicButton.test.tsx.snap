// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MagicButton > х┐лчЕзц╡ЛшпХ > ф╕НхРМх░║хп╕цМЙщТох┐лчЕз 1`] = `
<DocumentFragment>
  <button
    aria-describedby="test-id"
    class="magic-btn css-dev-only-do-not-override-yotw4n css-var-r2a magic-btn-default magic-btn-color-default magic-btn-variant-outlined magic-btn-sm acss-1i6z9ln"
    type="button"
  >
    <span>
      х░ПцМЙщТо
    </span>
  </button>
</DocumentFragment>
`;

exports[`MagicButton > х┐лчЕзц╡ЛшпХ > ф╕НхРМх░║хп╕цМЙщТох┐лчЕз 2`] = `
<DocumentFragment>
  <button
    aria-describedby="test-id"
    class="magic-btn css-dev-only-do-not-override-yotw4n css-var-r2e magic-btn-default magic-btn-color-default magic-btn-variant-outlined magic-btn-lg acss-1i6z9ln"
    type="button"
  >
    <span>
      хдзцМЙщТо
    </span>
  </button>
</DocumentFragment>
`;

exports[`MagicButton > х┐лчЕзц╡ЛшпХ > ф╕╗шжБцМЙщТох┐лчЕз 1`] = `
<DocumentFragment>
  <button
    aria-describedby="test-id"
    class="magic-btn css-dev-only-do-not-override-yotw4n css-var-r1e magic-btn-primary magic-btn-color-primary magic-btn-variant-solid acss-1i6z9ln"
    type="button"
  >
    <span>
      ф╕╗шжБцМЙщТо
    </span>
  </button>
</DocumentFragment>
`;

exports[`MagicButton > х┐лчЕзц╡ЛшпХ > хКаш╜╜чК╢цАБцМЙщТох┐лчЕз 1`] = `
<DocumentFragment>
  <button
    aria-describedby="test-id"
    class="magic-btn css-dev-only-do-not-override-yotw4n css-var-r22 magic-btn-default magic-btn-color-default magic-btn-variant-outlined magic-btn-loading acss-1i6z9ln"
    type="button"
  >
    <span
      class="magic-btn-icon magic-btn-loading-icon"
    >
      <span
        aria-label="loading"
        class="anticon anticon-loading anticon-spin"
        role="img"
      >
        <svg
          aria-hidden="true"
          data-icon="loading"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="0 0 1024 1024"
          width="1em"
        >
          <path
            d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
          />
        </svg>
      </span>
    </span>
    <span>
      хКаш╜╜цМЙщТо
    </span>
  </button>
</DocumentFragment>
`;

exports[`MagicButton > х┐лчЕзц╡ЛшпХ > х╕жхЫ╛цаЗцМЙщТох┐лчЕз 1`] = `
<DocumentFragment>
  <button
    aria-describedby="test-id"
    class="magic-btn css-dev-only-do-not-override-yotw4n css-var-r26 magic-btn-default magic-btn-color-default magic-btn-variant-outlined acss-1i6z9ln"
    type="button"
  >
    <span
      class="magic-btn-icon"
    >
      <span>
        ЁЯЪА
      </span>
    </span>
    <span>
      х╕жхЫ╛цаЗцМЙщТо
    </span>
  </button>
</DocumentFragment>
`;

exports[`MagicButton > х┐лчЕзц╡ЛшпХ > цЦЗцЬмцМЙщТох┐лчЕз 1`] = `
<DocumentFragment>
  <button
    aria-describedby="test-id"
    class="magic-btn css-dev-only-do-not-override-yotw4n css-var-r1m magic-btn-text magic-btn-color-default magic-btn-variant-text acss-1i6z9ln"
    type="button"
  >
    <span>
      цЦЗцЬмцМЙщТо
    </span>
  </button>
</DocumentFragment>
`;

exports[`MagicButton > х┐лчЕзц╡ЛшпХ > чжБчФичК╢цАБцМЙщТох┐лчЕз 1`] = `
<DocumentFragment>
  <button
    aria-describedby="test-id"
    class="magic-btn css-dev-only-do-not-override-yotw4n css-var-r1u magic-btn-default magic-btn-color-default magic-btn-variant-outlined acss-1i6z9ln"
    disabled=""
    type="button"
  >
    <span>
      чжБчФицМЙщТо
    </span>
  </button>
</DocumentFragment>
`;

exports[`MagicButton > х┐лчЕзц╡ЛшпХ > шЩЪч║┐цМЙщТох┐лчЕз 1`] = `
<DocumentFragment>
  <button
    aria-describedby="test-id"
    class="magic-btn css-dev-only-do-not-override-yotw4n css-var-r1i magic-btn-dashed magic-btn-color-default magic-btn-variant-dashed acss-1i6z9ln"
    type="button"
  >
    <span>
      шЩЪч║┐цМЙщТо
    </span>
  </button>
</DocumentFragment>
`;

exports[`MagicButton > х┐лчЕзц╡ЛшпХ > щУ╛цОецМЙщТох┐лчЕз 1`] = `
<DocumentFragment>
  <button
    aria-describedby="test-id"
    class="magic-btn css-dev-only-do-not-override-yotw4n css-var-r1q magic-btn-link magic-btn-color-link magic-btn-variant-link acss-1i6z9ln"
    type="button"
  >
    <span>
      щУ╛цОецМЙщТо
    </span>
  </button>
</DocumentFragment>
`;

exports[`MagicButton > х┐лчЕзц╡ЛшпХ > щЪРшЧПцМЙщТох┐лчЕз 1`] = `<DocumentFragment />`;

exports[`MagicButton > х┐лчЕзц╡ЛшпХ > щ╗ШшодцМЙщТох┐лчЕз 1`] = `
<DocumentFragment>
  <button
    aria-describedby="test-id"
    class="magic-btn css-dev-only-do-not-override-yotw4n css-var-r1a magic-btn-default magic-btn-color-default magic-btn-variant-outlined acss-1i6z9ln"
    type="button"
  >
    <span>
      щ╗ШшодцМЙщТо
    </span>
  </button>
</DocumentFragment>
`;
